File: src/me/anno/image/Image.java
package me.anno.image;

import me.anno.cache.data.ICacheData;
import me.anno.gpu.texture.Texture2D;
import me.anno.image.raw.BIImage;
import me.anno.image.raw.IntImage;
import me.anno.io.files.FileReference;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.awt.image.DataBuffer;
import java.awt.image.DataBufferInt;
import java.awt.image.Raster;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;

public abstract class Image implements ICacheData {

    int numChannels;
    public boolean hasAlphaChannel;

    public Image(int width, int height, int numChannels, boolean hasAlphaChannel) {
        this.numChannels = numChannels;
        this.hasAlphaChannel = hasAlphaChannel;
        this.width = width;
        this.height = height;
    }

    public int width, height;

    public int getIndex(int x, int y) {
        x = Math.max(Math.min(x, width - 1), 0);
        y = Math.max(Math.min(y, height - 1), 0);
        return x + y * width;
    }

    public int getNumChannels() {
        return numChannels;
    }

    public void setHasAlphaChannel(boolean alpha) {
        hasAlphaChannel = alpha;
    }

    public boolean hasAlphaChannel() {
        return hasAlphaChannel;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public IntImage createIntImage() {
        int width = getWidth();
        int height = getHeight();
        int[] data = new int[width * height];
        IntImage image = new IntImage(width, height, data, hasAlphaChannel);
        for (int i = 0, size = width * height; i < size; i++) {
            data[i] = getRGB(i);
        }
        return image;
    }

    public BIImage createBImage(int width, int height) {
        return new BIImage(createBufferedImage(width, height));
    }

    public BufferedImage createBufferedImage() {
        int width = getWidth();
        int height = getHeight();
        BufferedImage image = new BufferedImage(width, height, hasAlphaChannel ? 2 : 1);
        DataBufferInt dataBuffer = (DataBufferInt) image.getRaster().getDataBuffer();
        int[] data = dataBuffer.getData();
        for (int i = 0, size = width * height; i < size; i++) {
            data[i] = getRGB(i);
        }
        return image;
    }

    public abstract int getRGB(int index);

    public float getValueAt(float x, float y, int shift) {

        float xf = (float) Math.floor(x);
        float yf = (float) Math.floor(y);

        int xi = (int) xf;
        int yi = (int) yf;

        float fx = x - xf, gx = 1f - fx;
        float fy = y - yf, gy = 1f - fy;

        int c00, c01, c10, c11;
        int width = this.width;
        if (xi >= 0 && yi >= 0 && xi < width - 1 && yi < height - 1) {
            // safe
            int index = xi + yi * width;
            c00 = getRGB(index);
            c01 = getRGB(index + width);
            c10 = getRGB(index + 1);
            c11 = getRGB(index + 1 + width);
        } else {
            // border
            c00 = getSafeRGB(xi, yi);
            c01 = getSafeRGB(xi, yi + 1);
            c10 = getSafeRGB(xi + 1, yi);
            c11 = getSafeRGB(xi + 1, yi + 1);
        }

        c00 = (c00 >> shift) & 255;
        c01 = (c01 >> shift) & 255;
        c10 = (c10 >> shift) & 255;
        c11 = (c11 >> shift) & 255;

        float r0 = c00 * gy + fy * c01;
        float r1 = c10 * gy + fy * c11;
        return r0 * gx + fx * r1;

    }

    public int getRGB(int x, int y) {
        return getRGB(x + y * width);
    }

    public int getSafeRGB(int x, int y) {
        return getRGB(getIndex(x, y));
    }

    public void createTexture(Texture2D texture, boolean checkRedundancy) {
        texture.create(createBufferedImage(), true, true);
    }

    public BufferedImage createBufferedImage(int dstWidth, int dstHeight) {
        final int srcWidth = this.width;
        final int srcHeight = this.height;
        if (dstWidth > srcWidth) {
            dstHeight = dstHeight * srcWidth / dstWidth;
            dstWidth = srcWidth;
        }
        if (dstHeight > srcHeight) {
            dstWidth = dstWidth * srcHeight / dstHeight;
            dstHeight = srcHeight;
        }
        if (dstWidth == srcWidth && dstHeight == srcHeight) {
            return createBufferedImage();
        }
        final BufferedImage img = new BufferedImage(dstWidth, dstHeight, hasAlphaChannel ? 2 : 1);
        final DataBuffer buffer = img.getData().getDataBuffer();
        int srcY0 = 0;
        for (int dstY = 0, dstIndex = 0; dstY < dstHeight; dstY++) {
            final int srcY1 = ((dstY + 1) * srcHeight) / dstHeight;
            final int srcDY = srcY1 - srcY0;
            final int srcIndexY0 = srcY0 * srcWidth;
            for (int dstX = 0; dstX < dstWidth; dstX++, dstIndex++) {
                int srcX0 = (dstX * srcWidth) / dstWidth;
                int srcX1 = ((dstX + 1) * srcWidth) / dstWidth;
                // we could use better interpolation, but it shouldn't really matter
                int r = 0, g = 0, b = 0, a = 0;
                int srcIndexYI = srcIndexY0;
                // use interpolation
                for (int y0 = srcY0; y0 < srcY1; y0++) {
                    final int startIndex = srcX0 + srcIndexYI;
                    final int endIndex = srcX1 + srcIndexYI;
                    for (int i = startIndex; i < endIndex; i++) {
                        final int color = getRGB(i);
                        a += (color >> 24) & 255;
                        r += (color >> 16) & 255;
                        g += (color >> 8) & 255;
                        b += color & 255;
                    }
                    srcIndexYI += srcWidth;
                }
                final int count = (srcX1 - srcX0) * srcDY;
                if (count > 1) {
                    a /= count;
                    r /= count;
                    g /= count;
                    b /= count;
                }
                buffer.setElem(dstIndex, argb(a, r, g, b));
            }
            srcY0 = srcY1;
        }
        // update the image, otherwise the result is broken
        img.setData(Raster.createRaster(img.getRaster().getSampleModel(), buffer, new Point()));
        return img;
    }

    public void write(FileReference dst) throws IOException {
        String format = dst.getLcExtension();
        try (OutputStream out = dst.outputStream()) {
            write(out, format);
        }
    }

    public void write(OutputStream dst, String format) throws IOException {
        BufferedImage image = createBufferedImage();
        ImageIO.write(image, format, dst);
    }

    public static int argb(int a, int r, int g, int b) {
        return (a << 24) + (r << 16) + (g << 8) + b;
    }

    public static void createRGBFrom3StridedData(Texture2D texture, int width, int height, boolean checkRedundancy, byte[] data) {
        // add a padding for alpha, because OpenGL needs it that way
        ByteBuffer buffer = Texture2D.Companion.getBufferPool().get(width * height * 4, false);
        for (int j = 0, k = 0, l = width * height * 3; k < l; ) {
            buffer.put(j++, (byte) 255);// a
            buffer.put(j++, data[k++]);// r
            buffer.put(j++, data[k++]);// g
            buffer.put(j++, data[k++]);// b
        }
        texture.createRGB(buffer, checkRedundancy);
    }

    @Override
    public void destroy() {
    }

}

File: src/me/anno/image/Image.kt
package me.anno.image

import me.anno.cache.ICacheData
import me.anno.gpu.GFX
import me.anno.gpu.GPUTasks.addGPUTask
import me.anno.gpu.texture.Clamping
import me.anno.gpu.texture.Filtering
import me.anno.gpu.texture.ITexture2D
import me.anno.gpu.texture.Texture2D
import me.anno.image.bmp.BMPWriter
import me.anno.image.hdr.HDRWriter
import me.anno.image.raw.IFloatImage
import me.anno.image.raw.IntImage
import me.anno.io.files.FileReference
import me.anno.io.files.InvalidRef
import me.anno.io.files.inner.temporary.InnerTmpImageFile
import me.anno.maths.Maths.clamp
import me.anno.utils.Color.mixARGB22d
import me.anno.utils.Color.toARGB
import me.anno.utils.InternalAPI
import me.anno.utils.Logging.hash32
import me.anno.utils.async.Callback
import me.anno.utils.callbacks.I2U
import me.anno.utils.hpc.WorkSplitter
import org.joml.Vector4f
import java.io.OutputStream
import kotlin.math.floor

abstract class Image(
    open var width: Int,
    open var height: Int,
    var numChannels: Int,
    var hasAlphaChannel: Boolean,
    var offset: Int,
    var stride: Int,
) : ICacheData {

    constructor(width: Int, height: Int, numChannels: Int, hasAlphaChannel: Boolean) :
            this(width, height, numChannels, hasAlphaChannel, 0, width)

    override fun toString(): String {
        return "${this::class.simpleName}@${hash32(this)}[$width x $height x $numChannels${if (hasAlphaChannel) ", alpha" else ""}]"
    }

    open fun getIndex(x: Int, y: Int): Int {
        val xi = clamp(x, 0, width - 1)
        val yi = clamp(y, 0, height - 1)
        return offset + xi + yi * stride
    }

    open fun asIntImage(): IntImage {
        return cloneToIntImage()
    }

    open fun cloneToIntImage(): IntImage {
        val width = width
        val height = height
        val data = IntArray(width * height)
        var i = 0
        for (y in 0 until height) {
            for (x in 0 until width) {
                data[i++] = getRGB(x, y)
            }
        }
        return IntImage(width, height, data, hasAlphaChannel)
    }

    /**
     * get the argb-color at that pixel index;
     * please use getIndex() or getRGB(x,y) instead of calculating the index yourself
     * */
    abstract fun getRGB(index: Int): Int

    /**
     * return r/g/b/a value at shift(r=16,g=8,b=0,a=24) as a float [0,255]
     * */
    fun getValueAt(x: Float, y: Float, shift: Int): Float {
        val xf = floor(x)
        val yf = floor(y)
        val xi = xf.toInt()
        val yi = yf.toInt()
        val fx = x - xf
        val gx = 1f - fx
        val fy = y - yf
        val gy = 1f - fy
        val c00 = getSafeRGB(xi, yi).shr(shift).and(255)
        val c01 = getSafeRGB(xi, yi + 1).shr(shift).and(255)
        val c10 = getSafeRGB(xi + 1, yi).shr(shift).and(255)
        val c11 = getSafeRGB(xi + 1, yi + 1).shr(shift).and(255)
        val r0 = c00 * gy + fy * c01
        val r1 = c10 * gy + fy * c11
        return r0 * gx + fx * r1
    }

    /**
     * returns argb value at these coordinates;
     * might crash if x or y are out of bounds
     * */
    fun getRGB(x: Int, y: Int): Int {
        return getRGB(getIndex(x, y))
    }

    fun setRGB(x: Int, y: Int, value: Int): Boolean {
        val validCoords = x in 0 until width && y in 0 until height
        if (validCoords) setRGB(getIndex(x, y), value)
        return validCoords
    }

    abstract fun setRGB(index: Int, value: Int)

    fun setRGB(x: Int, y: Int, value: Vector4f): Boolean {
        val validCoords = x in 0 until width && y in 0 until height
        if (validCoords) setRGB(getIndex(x, y), value)
        return validCoords
    }

    open fun setRGB(index: Int, value: Vector4f) {
        setRGB(index, value.toARGB())
    }

    fun sampleRGB(x: Float, y: Float, filtering: Filtering, clamping: Clamping): Int {
        return when (filtering) {
            Filtering.TRULY_NEAREST,
            Filtering.NEAREST -> getRGB(
                clamping.apply(floor(x).toInt(), width),
                clamping.apply(floor(y).toInt(), height)
            )
            Filtering.TRULY_LINEAR,
            Filtering.LINEAR -> {
                val xf = floor(x)
                val xi = xf.toInt()
                val fx = x - xf
                val yf = floor(y)
                val yi = yf.toInt()
                val fy = y - yf
                val width = width
                val height = height
                val x0 = clamping.apply(xi, width)
                val x1 = clamping.apply(xi + 1, width)
                val y0 = clamping.apply(yi, height)
                val y1 = clamping.apply(yi + 1, height)
                mixARGB22d(
                    getRGB(x0, y0), getRGB(x0, y1),
                    getRGB(x1, y0), getRGB(x1, y1),
                    fx, fy
                )
            }
        }
    }

    /**
     * returns argb value at these coordinates;
     * clamped access, if x or y are out of bounds
     * */
    fun getSafeRGB(x: Int, y: Int): Int {
        val xi = clamp(x, 0, width - 1)
        val yi = clamp(y, 0, height - 1)
        return getRGB(xi, yi)
    }

    fun createTexture(texture: Texture2D, checkRedundancy: Boolean, callback: Callback<ITexture2D>) {
        if (GFX.isGFXThread()) {
            createTextureImpl(texture, checkRedundancy, callback)
        } else {
            addGPUTask("Image.createTexture", width, height) {
                createTextureImpl(texture, checkRedundancy, callback)
            }
        }
    }

    open fun createTextureImpl(texture: Texture2D, checkRedundancy: Boolean, callback: Callback<ITexture2D>) {
        texture.create(asIntImage(), sync = true, checkRedundancy = true, callback)
    }

    open fun resized(dstWidth: Int, dstHeight: Int, allowUpscaling: Boolean): Image {
        return ImageResizing.resized(this, dstWidth, dstHeight, allowUpscaling)
    }

    fun resampled(dstWidth: Int, dstHeight: Int): Image {
        return ResampledImage(this, dstWidth, dstHeight)
    }

    open fun write(dst: FileReference, quality: Float = 0.9f) {
        val format = dst.lcExtension
        dst.outputStream().use { out -> write(out, format, quality) }
    }

    fun write(dst: OutputStream, format: String, quality: Float = 0.9f) {
        writeImageImpl.write(this, dst, format, quality)
    }

    override fun destroy() {}

    /**
     * for debugging and easier seeing pixels: nearest filtering, no interpolation
     * */
    open fun scaleUp(sx: Int, sy: Int = sx): Image {
        return ScaledImage(this, sx, sy)
    }

    open fun split(numTilesX: Int, numTilesY: Int): List<Image> {
        return List(numTilesX * numTilesY) { tileId ->
            val ix = tileId % numTilesX
            val iy = tileId / numTilesX
            val x0 = WorkSplitter.partition(ix, width, numTilesX)
            val x1 = WorkSplitter.partition(ix + 1, width, numTilesX)
            val y0 = WorkSplitter.partition(iy, height, numTilesY)
            val y1 = WorkSplitter.partition(iy + 1, height, numTilesY)
            cropped(x0, y0, x1 - x0, y1 - y0)
        }
    }

    open fun cropped(x0: Int, y0: Int, w0: Int, h0: Int): Image {
        return CroppedImage(this, x0, y0, w0, h0)
    }

    /**
     * execution order isn't guaranteed!
     * */
    fun forEachPixel(callback: I2U) {
        for (y in 0 until height) {
            for (x in 0 until width) {
                callback.call(x, y)
            }
        }
    }

    fun flipY() {
        offset = getIndex(0, height - 1)
        stride = -stride
    }

    var ref: FileReference = InvalidRef
        get() {
            if (field == InvalidRef) {
                field = InnerTmpImageFile(this)
            }
            return field
        }

    fun sizeGuess(): Long {
        return width.toLong() * height.toLong() * numChannels.toLong()
    }

    companion object {

        @InternalAPI
        var writeImageImpl = ImageStreamWriter { img, out, format, _ ->
            if ("hdr".equals(format, true)) {
                writeHDR(img, out)
            } else {
                out.write(BMPWriter.createBMP(img))
            }
        }

        fun writeHDR(image: Image, output: OutputStream) {
            if (image is IFloatImage && image.numChannels >= 3) {
                val floatImage = image.toFloatImage(false)
                HDRWriter.writeHDR(
                    floatImage.width, floatImage.height, floatImage.stride,
                    floatImage.data, output
                )
            } else {
                output.write(BMPWriter.createBMP(image))
            }
        }
    }
}